### ISSUE TITLE
Gluster-block command not found in glusterfs pods

### ISSUE THREAD

**Issue by ahmadou** [openshift/origin#19451]
[provide a description of the issue]

I want to be able to create pvc on a glusterfs block storage.
I followed this tutorial to add a glusterblock storage class

https://access.redhat.com/documentation/en-us/red_hat_gluster_storage/3.3/html/container-native_storage_for_openshift_container_platform/block_storage.

I'm able to create the storageclass but the pvc then cannot be created.

##### Version

oc v3.7.1+c2ce2c0-1
kubernetes v1.7.6+a08f5eeb62
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://masta:8443
openshift v3.7.1+c2ce2c0-1
kubernetes v1.7.6+a08f5eeb62

##### Steps To Reproduce

Follow the steps in the tutorial to setup the storageclass.
Then create a pvc.

Alternatively : 

Log to heketi pod and try to create a block volume.

##### Current Result
The class is created but pvc fails with error : 

`Failed to provision volume with StorageClass "glusterfs-block-storage": failed to create volume: [heketi] failed to create volume: Unable to execute command on glusterfs-storage-j549n: /bin/bash: gluster-block: command not found`

##### Expected Result

Pvc are created according to documentation.

##### Additional Information

I also tried to manually use the heketi-cli to add a block volume and it fails with the same error.
The gluster-block command is indeed absent from the glusterfs containers.
Am i missing something here ?

**Comment by jwforres** [comment]
@openshift/sig-storage

**Comment by ahmadou** [comment]
If i may, i made it work by ditching the containerized gluster install and setting up an external glusterfs cluster.
I then installed the cluster with the supported external cluster setting.
With a bit of tweaking it is working now and i can provision volumes with gluster block.

**Comment by ahmadou** [comment]
First pain is to have all the necessary librairies up and running.
Fortunately I made an ansible playbook to install and configure the gluster cluster with gluster block enabled
```
--
- hosts: all
  remote_user: ansibleuser
  become: yes
  become_method: sudo
  tasks:
    - name: 1. install htop
      yum: name=htop state=present

    - name: 2 install gusterfs centos release repo
      yum: name=centos-release-gluster state=present

    - name: 3. install gluster fs packages
      yum: 
        state: present
        name: 
          - glusterfs-server 
          - parted 
          - lvm2
          - xfsprogs
          - glusterfs
          - glusterfs-fuse
          - iscsi-initiator-utils 
          - device-mapper-multipath
          - autoconf
          - automake
          - libtool
          - libuuid-devel
          - json-c-devel
          - glusterfs-api-devel
          - targetcli
          - make

    - name: 4. Build tcmu runner
      shell: cd /tmp && rm -rf tcmu-runner && git clone https://github.com/open-iscsi/tcmu-runner && cd tcmu-runner && ./extra/install_dep.sh  && cmake -DSUPPORT_SYSTEMD=ON -DCMAKE_INSTALL_PREFIX=/usr . && make && make install

    - name: 5. Build gluster-block  package
      shell: cd /tmp && rm -rf gluster-block && git clone https://github.com/gluster/gluster-block.git && cd gluster-block/ && ./autogen.sh && ./configure && make -j install

    - name: 6. Ensure services are running
      service:
        state: restarted
        name: '{{ item }}'
        enabled: yes
      with_items:
        - tcmu-runner
        - glusterd
        - gluster-blockd
        - NetworkManager

```


This one will install and configure heketi

```
---
- hosts: all
  remote_user: ansibleuser
  become: yes
  become_method: sudo
  tasks:

    - name: 1 install gusterfs centos release repo
      yum: name=centos-release-gluster state=present

    - name: 2. install heketi package
      yum: 
        state: present
        name: 
          - heketi
          - heketi-client 

    - name: 3. Copy heketi private key. (you have to distribute the public key to the nodes yourself)
      copy:
        src: ../customKeys/heketi_key
        dest: /etc/heketi/heketi_key
        mode: 0644
        owner: heketi
        group: heketi

    - name: 4. Copy heketi configuration file
      copy:
        src: ../customFiles/heketi.json
        dest: /etc/heketi/heketi.json
        mode: 0644
        owner: heketi
        group: heketi


    - name: 4. Copy heketi topology file
      copy:
        src: ../customFiles/heketi-topology.json
        dest: /etc/heketi/heketi-topology.json
        mode: 0644
        owner: heketi
        group: heketi

    - name: 5. Ensure services are running
      service:
        state: restarted
        name: '{{ item }}'
        enabled: yes
      with_items:
        - heketi

```


Here is a sample heketi.json

```
{
  "_port_comment": "Heketi Server Port Number",
  "port": "8080",

  "_use_auth": "Enable JWT authorization. Please enable for deployment",
  "use_auth": true,

  "_jwt": "Private keys for access",
  "jwt": {
    "_admin": "Admin has access to all APIs",
    "admin": {
      "key": "<password admin>"
    },
    "_user": "User only has access to /volumes endpoint",
    "user": {
      "key": "<password user>"
    }
  },

  "_glusterfs_comment": "GlusterFS Configuration",
  "glusterfs": {
    "_executor_comment": [
      "Execute plugin. Possible choices: mock, ssh",
      "mock: This setting is used for testing and development.",
      "      It will not send commands to any node.",
      "ssh:  This setting will notify Heketi to ssh to the nodes.",
      "      It will need the values in sshexec to be configured.",
      "kubernetes: Communicate with GlusterFS containers over",
      "            Kubernetes exec api."
    ],
    "executor": "ssh",

    "_sshexec_comment": "SSH username and private key file information",
    "sshexec": {
      "keyfile": "/etc/heketi/heketi_key",
      "user": "root"
    },

    "_kubeexec_comment": "Kubernetes configuration",
    "kubeexec": {
      "host" :"https://kubernetes.host:8443",
      "cert" : "/path/to/crt.file",
      "insecure": false,
      "user": "kubernetes username",
      "password": "password for kubernetes user",
      "namespace": "OpenShift project or Kubernetes namespace",
      "fstab": "Optional: Specify fstab file on node.  Default is /etc/fstab"
    },

    "_db_comment": "Database file name",
    "db": "/var/lib/heketi/heketi.db",

    "_loglevel_comment": [
      "Set log level. Choices are:",
      "  none, critical, error, warning, info, debug",
      "Default is warning"
    ],
    "loglevel" : "debug"
  }
}

```

Sample topology file
```
{
  "clusters": [
    {
      "nodes": [
        {
          "node": {
            "hostnames": {
              "manage": [
                "host1"
              ],
              "storage": [
                "1.1.1.1"
              ]
            },
            "zone": 1
          },
          "devices": [
            "/dev/sdc",
            "/dev/sdd"      

          ]
        },
       {
          "node": {
            "hostnames": {
              "manage": [
                "host2"
              ],
              "storage": [
                "1.1.1.2"
              ]
            },
            "zone": 1
          },
          "devices": [
            "/dev/sdc",
            "/dev/sdd"      

          ]
        }
      ]
    }
  ]
}

```

You'll have to edit your topology file according to your hosts and devices.
Then you have to load it to create the cluster via heketi.
Then you have to manually create special volumes that will house the block vloumes
Use this command :
`heketi-cli volume create --size=200 --block
`
You'll have to always ensure yourself that there is  enough block volumes of this kind and that they have free space.
Openshift won't create them for you if you go this route.

These "block volumes" will be used for your gluster block pvc and are different from the gluster volumes created by your standard  gluster pvc (It took me some time to figure this one out).

After that you can follow the tutorial : 

https://access.redhat.com/documentation/en-us/red_hat_gluster_storage/3.3/html/container-native_storage_for_openshift_container_platform/block_storage.

It is normal that the url is the same. Heketi still manages the cluster.

Make sure to still use the glusterfs openshift playbook but set these parameters :+1: 


```
openshift_storage_glusterfs_namespace=glusterfs
openshift_storage_glusterfs_is_native=false
openshift_storage_glusterfs_storageclass=true
openshift_storage_glusterfs_storageclass_default=true
openshift_storage_glusterfs_block_storageclass=true
openshift_storage_glusterfs_heketi_is_native=false
openshift_storage_glusterfs_heketi_admin_key=<same key used in your conf file>
openshift_storage_glusterfs_heketi_user_key=<same key used in your conf file>
openshift_storage_glusterfs_heketi_topology_load=false
openshift_storage_glusterfs_heketi_url=<your heketi host>
openshift_storage_glusterfs_heketi_port=8080
```

This will deploy the glusterblock storage provisioner pod which is needed so that your pvc can be processed. Otherwise they'll hang indefinitely.

Normally it should work. Hope this helps.

**Comment by lentzi90** [comment]
Thank you for the help! I was able to get it working following [this](https://docs.openshift.org/3.9/install_config/persistent_storage/persistent_storage_glusterfs.html#install-example-full-external) guide when first installing
- NetworkManager
- centos-release-gluster40
- glusterfs-server
- gluster-block

on all gluster nodes, as well as starting and enabling `NetworkManager`, `glusterd` and `gluster-blockd`.

I also changed a few settings from the guide, namely:
- added `openshift_metrics_image_version=v3.9` since there is no v3.9.0 tag for the metrics images.
- `openshift_storage_glusterfs_storageclass_default=true` to allow ansible service broker to start with a PVC.
- `openshift_storage_glusterfs_registry_block_storageclass_default=false` since the other storage class was default.

After the initial deployment of the cluster but before adding logging and metrics, change the default storage class to `glusterfs-registry-block`.
```
oc patch storageclass glusterfs-storage -p '{"metadata": {"annotations": {"storageclass.kubernetes.io/is-default-class": "false"}}}'
oc patch storageclass glusterfs-registry-block -p '{"metadata": {"annotations": {"storageclass.kubernetes.io/is-default-class": "true"}}}'
```
